### PR TITLE
home-environment: add Nixpkgs release version check

### PR DESCRIPTION
**Pull request by rycee**
### Description

This adds a warning when a release version mismatch is detected between Home Manager and Nixpkgs.

### Checklist

- [x] Change is backwards compatible.

- [X] Code formatted with `./format`.

- [X] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

### CODE REVIEW

File: modules/home-environment.nix [hunk 1]
```diff
@@ -404,6 +419,31 @@ in
       }
     ];
 
+    warnings =
+      let
+        hmRelease = fileContents ../.release;
+        nixpkgsRelease = pkgs.lib.trivial.release;
```
Comment (berbiche) [inline modules/home-environment.nix:425]: What's the difference with `pkgs.lib.version`?

Comment (rycee) [reply modules/home-environment.nix:425]: The `pkgs.lib.version` value is the same as `pkgs.lib.trivial.version`, for example, "21.05.git.e9148dc1c30" on my system.

The `pkgs.lib.trivial.release` value, on the other hand, is simply "21.05".

Comment (berbiche) [reply modules/home-environment.nix:425]: IIRC the values are different with a nixpkgs flake. I'll have to check

Comment (rycee) [reply modules/home-environment.nix:425]: Ah, seems odd to me if those would differ in a flake context but it would certainly be good to verify, thanks!

Comment (berbiche) [reply modules/home-environment.nix:425]:
``` console
$ nix eval --impure --expr '(builtins.getFlake "nixpkgs").lib.version'
"21.11pre-git"

$ nix eval --impure --expr '(builtins.getFlake "nixpkgs").lib.trivial.release'
"21.11"
```

It only affects `lib.version` so it won't be a problem!

File: modules/home-environment.nix [hunk 2]
```diff
@@ -404,6 +419,31 @@ in
...
+      in
+        optional releaseMismatch ''
+          You are using
+
+            Home Manager version ${hmRelease} and
+            Nixpkgs version ${nixpkgsRelease}.
+
+          Using mismatched versions is likely to cause errors and unexpected
+          behavior. It is therefore highly recommended to use a release of Home
+          Manager than correspond with your chosen release of Nixpkgs.
```
Comment (berbiche) [inline modules/home-environment.nix:438]:
```suggestion
          Manager that corresponds with your chosen release of Nixpkgs.
```

File: modules/home-environment.nix [hunk 3]
```diff
@@ -390,6 +390,21 @@ in
...
     };
+
+    home.enableNixpkgsReleaseCheck = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        Determines whether to check for release version mismatch between Home
+        Manager and Nixpkgs. Using mismatched versions is likely to cause errors
+        and unexpected behavior. It is therefore highly recommended to use a
+        release of Home Manager than correspond with your chosen release of
```
Comment (berbiche) [inline modules/home-environment.nix:401]:
```suggestion
        release of Home Manager that corresponds with your chosen release of
```

**Review by berbiche — APPROVED** [review]

**Pull request merged**